### PR TITLE
issue  #196 Upgrade for hapi v17?

### PR DESCRIPTION
**Pull request by lancelot316**
I updated as best I could the problem and solution files to hapi v17. I was unable to update the upload tutorial, because I couldn't figure out how to make it work without the reply callback.

I have seen a several different ways to configure the server and call server.start. 
I used an async IIFE instead of a named function and a separate call to that function, because it seemed a little cleaner to me.

### CODE REVIEW

File: exercises/auth/solution/solution.js [hunk 1]
```diff
@@ -1,35 +1,39 @@
...
-server.register(Auth, (err) => {
-    server.auth.strategy('simple', 'basic', { validateFunc: validate });
-    server.route({
-        method: 'GET',
-        path: '/',
-        config: {
-            auth: 'simple',
-            handler: (request, reply) => {
-                reply();
+(async () => {
```
Comment (devinivy) [inline exercises/auth/solution/solution.js]: I put some time into thinking about what the entrypoint to a server should look like, and here's what I came-up with: https://github.com/devinivy/boilerplate-api/blob/pal/server/index.js .  The main idea is to add a couple lines to throw unhandled rejections.

This was also a good conversation about it: https://github.com/hapijs/discuss/issues/570

**Comment by fiveisprime** [comment]
I'm unable to run this as is, looks like the dependency versions haven't been updated in package.json. Can you commit the changes to package.json?

**Comment by fiveisprime** [comment]
The proxies exercise needs updated here: https://github.com/lancelot316/makemehapi/blob/master/exercises/proxies/exercise.js#L36-L48

File: exercises/helping/solution/solution.js [hunk 2]
```diff
@@ -1,34 +1,37 @@
...
 
-server.register(Vision, (err) => {
-    if (err) throw err;
-});
+        server.views({
+            engines: { 
+                html: Handlebars
+            },
+            path: Path.join(__dirname, 'templates'),
+            helpersPath: 'helpers'
```
Comment (fiveisprime) [inline exercises/helping/solution/solution.js]: `helpersPath` needs `Path.join(__dirname, 'templates')` as well

**Comment by lancelot316** [comment]
I made the changes you mentioned. I think I got everything, except the uploads exercise I'm still learning the framework, but I figured as I was learning Hapi I could try and contribute back what I have learned so far.

File: exercises/proxies/exercise.js [hunk 3]
```diff
@@ -32,21 +31,29 @@ exercise.addSetup(function (mode, callback) {
...
+            var server = new Hapi.Server({
+                host: 'localhost',
+                port: 65535
+            });
+
+            server.route({
+                method: 'GET',
+                path: '/proxy',
+                handler: function (request, reply) {
+                    reply(exercise.__('greeting'));
```
Comment (fiveisprime) [inline exercises/proxies/exercise.js]: Should be `return exercise.__('greeting');`

**Comment by fiveisprime** [comment]
Thanks again! There's one more change needed in the proxies exercise to _return_ the response rather than use the old `reply` method.

Also, this is what I came up with for the uploads solution, you're welcome to build on/use this: https://gist.github.com/fiveisprime/99605395e87f6b3d9af87f49316fe43b

I'm not 100% on if that's optimal; I had to return a promise to allow for the stream processing. There may be a better way to handle that and there isn't any error handling for the rejection.

**Pull request merged**